pkg/executor/fscache/poolcache.go

```diff
@@ -156,14 +168,55 @@ func (c *PoolCache) service() {
 				continue
 			}
 
-			// concurrency should not be set to zero and
-			//sum of specialization in progress and specialized pods should be less then req.concurrency
-			if req.concurrency > 0 && (specializationInProgress+len(funcSvcGroup.svcs)) >= req.concurrency {
-				resp.error = ferror.MakeError(ferror.ErrorTooManyRequests, fmt.Sprintf("function '%s' concurrency '%d' limit reached.", req.function, req.concurrency))
-			} else {
-				funcSvcGroup.svcWaiting++
-				resp.error = ferror.MakeError(ferror.ErrorNotFound, fmt.Sprintf("function '%s' all functions are busy", req.function))
+			numPodsSpecialized := len(funcSvcGroup.svcs)
+
+			// we're either not specializing a pod, or we are and none of them have room for another request
+			// check if we can specialize another pod
+			if req.concurrency > 0 && numPodsSpecialized+numPodsSpecializing >= req.concurrency {
+				// we can't specialize another pod because our specialized pods + specializing pods is >= this function's concurrency limit
+				// check if we have any pods that are already specialized
+				if numPodsSpecialized > 0 {
+					// we have specialized pod(s), just pick one at random and use it for this request
+					svc := randomSvc(funcSvcGroup.svcs)
```

Comment:
Curious what you all think about this scenario:

1. One specialized pod
2. A massive burst of requests, going well beyond the max requests
3. All requests that spill over the max requests go to that one specialized pod because the others are specializing

Could we do more harm than good by having one pod take those request? Should we consider an upper bound where we start putting these requests on the queue for a specializing pod to become available?

Reply:
That's a good point, how about this instead:

```go
if numPodsSpecializing > 0 {
	// we still have pods that are specializing
	// increment the number of requests waiting for a pod to finish specializing and push this request onto the queue
} else {
	// we have the maximum amount of specialized pods, just pick one at random and use it for this request
}
```

☝️ This queues the requests for the specializing pods instead of the specialized one(s). We already have logic [below](https://github.com/gadget-inc/fission/pull/19/files/63ef57691b85cd76ec0e3c1de1abbfb98cd07244#diff-c0519a1706a14c015ac4471698f6033c5d627c1ba15fa53f85dec3086b3b18ecR252) to randomly distribute the queue among all the pods once we've specialized the last one.

Reply:
Yeah, I think in the absolute worst-case scenario outlined above, I'd rather have those requests distributed, or at least some of them, distributed across the specializing pods to not overwhelm the one specialized pod. Some questions that come to my mind:

* Do we have any sense of what would be overwhelming? I know node could definitely take on a ton of requests, but eventually it would topple over.
* Latency-wise, anything we can measure? Would it be better to wait for specialization, or have the specialized pod slowly work through the requests?
* If specialization fails, we keep things in the queue. Does that mean, for now, it would be best to send it at the specialized pod instead and follow up where we pop things off the queue when an existing specialized pod can serve more requests, i.e. when `markAvailable`?

I'm not sure what's best, but I'll let you make that call!